data_steward/cdr_cleaner/cleaning_rules/unit_normalization.py

```diff
@@ -0,0 +1,283 @@
+        """
+        desc = 'Units for labs/measurements will be normalized using unit_mapping lookup table.'
+        super().__init__(issue_numbers=['DC414', 'DC700'],
+                         description=desc,
+                         affected_datasets=[cdr_consts.DEID_CLEAN],
+                         affected_tables=['measurement'],
+                         project_id=project_id,
+                         dataset_id=dataset_id,
+                         sandbox_dataset_id=sandbox_dataset_id)
+
+    def setup_rule(self, client):
+        """
+        Load required resources prior to executing cleaning rule queries.
+
+        Method to run data upload options before executing the first cleaning
+        rule of a class.  For example, if your class requires loading a static
+        table, that load operation should be defined here.  It SHOULD NOT BE
+        defined as part of get_query_specs().
+        :param client:
+        :return:
+        """
+
+        # creating _unit_mapping table
+        unit_mapping_table = f'{self.get_project_id()}.{self.get_dataset_id()}.{UNIT_MAPPING_TABLE}'
+        bq.create_tables(
+            client,
+            self.get_project_id(),
+            [unit_mapping_table],
+        )
+        # Uploading data to _unit_mapping table
+        unit_mappings_csv_path = os.path.join(resources.resource_files_path,
+                                              UNIT_MAPPING_FILE)
+        job = bq.upload_csv_data_to_bq_table(client, self.get_dataset_id(),
+                                             UNIT_MAPPING_TABLE,
+                                             unit_mappings_csv_path,
+                                             UNIT_MAPPING_TABLE_DISPOSITION)
+        LOGGER.info(
+            f"Created {self.get_dataset_id()}.{UNIT_MAPPING_TABLE} and loaded data from {unit_mappings_csv_path}"
+        )
+        try:
+            job.result()
+        except BadRequest as e:
+            for e in job.errors:
+                print(f'ERROR: {e["message"]}')
+
+    def get_query_specs(self):
+        """
+        :return:
+        """
+        sandbox_query = dict()
+        sandbox_query[cdr_consts.QUERY] = SANDBOX_UNITS_QUERY.render(
+            project_id=self.get_project_id(),
+            sandbox_dataset=self.get_sandbox_dataset_id(),
+            intermediary_table=self.get_sandbox_tablenames()[0],
+            dataset_id=self.get_dataset_id(),
+            unit_table_name=UNIT_MAPPING_TABLE,
+            measurement_table=MEASUREMENT)
+
+        update_query = dict()
+        update_query[cdr_consts.QUERY] = UNIT_NORMALIZATION_QUERY.render(
+            project_id=self.get_project_id(),
+            dataset_id=self.get_dataset_id(),
+            unit_table_name=UNIT_MAPPING_TABLE,
+            measurement_table=MEASUREMENT)
+        update_query[cdr_consts.DESTINATION_TABLE] = MEASUREMENT
+        update_query[cdr_consts.DISPOSITION] = bq_consts.WRITE_TRUNCATE
```

Comment:
`WRITE_TRUNCATE` would wipe out the schema, right? Have you ever looked into instead using `CREATE OR REPLACE TABLE ... AS query_statement` ([source](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#creating_a_clustered_table_from_the_result_of_a_query))?

```sql 
CREATE OR REPLACE TABLE source_dataset.measurement(
  measurement_id INT64 NOT NULL OPTIONS(description='...')
  /* ... */
) AS 
SELECT m.* 
FROM source_dataset.measurement m
WHERE /* ... */
```

I'm not suggesting we do this here, but it is something I'm trying out somewhere else and I wanted to know if you'd looked into it already.

Reply:
I looked into this solution and thought this would change a lot of our cleaning process and wanted to discuss in the ENG-Sync before actually making that decision.

Reply:
Ok, just checking. Closing this thread.